### PR TITLE
Feature: Crypto payments.

### PR DESCRIPTION
**Pull request by Danziger**

### CODE REVIEW

File: app/src/lib/components/payments/PaymentDetailsItem/Fragment/PaymentDetailsFragment.tsx [hunk 1]
```diff
@@ -22,10 +24,23 @@ export const PaymentDetailsFragment: React.FC<PaymentDetailsFragmentProps> = ({
...
+        <Typography>{ ACH_MASK_PREFIX } { savedPaymentMethod.accountNumber }</Typography>
+      </>
+    );
+  }
+
+  if (savedPaymentMethod.type === "Crypto") {
+    return (
+      <>
+        <Icon path={ mdiEthereum } size="24px" />
+        <Typography>{ savedPaymentMethod.id }</Typography>
```
Comment (Danziger) [inline app/src/lib/components/payments/PaymentDetailsItem/Fragment/PaymentDetailsFragment.tsx:40]: We don't really have anything meaningful to display here (list of saved payment methods).

File: app/src/lib/components/payments/PurchaseConfirmationBillingDetails/PurchaseConfirmationBillingDetails.utils.ts [hunk 2]
```diff
@@ -25,14 +25,17 @@ export function getFormattedPaymentMethod(paymentMethodInfo: PaymentMethod | Sav
...
       paymentType = "ACH";
       displayValue = `${ ACH_MASK_PREFIX } ${ paymentMethodInfo.accountNumber }`;
-    } else {
+    } else if (paymentMethodInfo.type === "CreditCard") {
       paymentType = "CreditCard";
       displayValue = `${ CREDIT_CARD_MASK_PREFIX } ${ paymentMethodInfo.last4Digit }`;
       network = paymentMethodInfo.network;
+    } else if (paymentMethodInfo.type === "Crypto") {
+      paymentType = "Crypto";
+      displayValue = `${ paymentMethodInfo.id }`;
```
Comment (Danziger) [inline app/src/lib/components/payments/PurchaseConfirmationBillingDetails/PurchaseConfirmationBillingDetails.utils.ts:34]: We don't really have anything meaningful to display here (list of saved payment methods).

File: app/src/lib/domain/circle/circle.interfaces.ts [hunk 3]
```diff
@@ -23,7 +23,15 @@ export interface RawSavedPaymentMethodACH {
   metadata: AchMetadata;
 }
 
-export type RawSavedPaymentMethod = (RawSavedPaymentMethodCC | RawSavedPaymentMethodACH) & { type: PaymentType };
+export interface RawSavedPaymentMethodCrypto {
+  id: string;
+  type: "Crypto";
+  status: PaymentMethodStatus;
+  billingDetails: undefined;
```
Comment (Danziger) [inline app/src/lib/domain/circle/circle.interfaces.ts:30]: @Dmckinney821 It is a bit weird that all payment methods, when created, need to include `billingDetails` and `metadata`, while `Crypto` doesn't have them and they are sent to `createPayment` instead. This is actually causing issues with all the logic in the frontend to show saved addresses and payment methods.

File: app/src/lib/hooks/useCreatePaymentMethod.ts [hunk 4]
```diff
@@ -90,11 +78,20 @@ export function useCreatePaymentMethod({
...
             },
           },
         },
       });
+    } else if (paymentInfo.type === PaymentType.Crypto) {
+      createPaymentMethodPromise = createPaymentMethod({
+        variables: {
+          orgID,
+          input: {
+            paymentType: PaymentType.Crypto,
```
Comment (Danziger) [inline app/src/lib/hooks/useCreatePaymentMethod.ts:91]: Ideally, both `billingDetails` and metadata should be sent here.

If you want to buy NFT 1 with address A and NFT 2 with address B, both using Crypto, two separated Crypto payment methods should be created, one for each address.

File: app/src/lib/hooks/useFullPayment.ts [hunk 5]
```diff
@@ -173,10 +181,13 @@ export function useFullPayment({
     }
 
     const metadata: CreatePaymentMetadataInput = destinationAddress ? { destinationAddress } : { };
+
     if (discountCodeID) {
       metadata.discountCodeID = discountCodeID;
     }
 
+    const isCrypto = true;
```
Comment (Danziger) [inline app/src/lib/hooks/useFullPayment.ts:189]: TODO: Finish implementing this.

File: app/src/lib/hooks/useFullPayment.ts [hunk 6]
```diff
@@ -200,6 +211,21 @@ export function useFullPayment({
...
         encryptedData: encryptedCardData,
       };
+    } else if (isCrypto) {
+      const currentURL = getUrlWithoutParams();
+      const billingDetails = selectedBillingInfoData ? billingInfoToBillingDetails<CryptoBillingDetails>(selectedBillingInfoData, "Crypto") : undefined;
+
+      metadata.cryptoData = {
+        name: checkoutItems[0].name,
+        description: checkoutItems[0].name,
+        billingDetails,
```
Comment (Danziger) [inline app/src/lib/hooks/useFullPayment.ts:221]: This doesn't belong here, but in `createPaymentMethod`.

File: app/src/lib/hooks/useFullPayment.ts [hunk 7]
```diff
@@ -200,6 +211,21 @@ export function useFullPayment({
...
+        redirectURL: currentURL,
+        cancelURL: currentURL,
+        localPrice: {
```
Comment (Danziger) [inline app/src/lib/hooks/useFullPayment.ts:224]: Why is this needed? The backend should get this data from the invoice as it also needs to account for applied discounts, which we can do on the frontend but it is a bit cumbersome and unnecessary anyway as the backend would have to check this is valid anyway.

File: app/src/lib/components/public/CheckoutOverlay/CheckoutOverlay.tsx [hunk 8]
```diff
@@ -491,6 +491,8 @@ export const PUICheckoutOverlay: React.FC<PUICheckoutOverlayProps> = ({
       if (typeof billingInfo === "string" && typeof paymentInfo === "string") return { ...prevSelectedPaymentMethod, cvv: "" };
 
       // To find the saved payment method(s) that was/were last created:
+
+      // TODO: This should be sorted by date which doesn't seem to be the case when different payment types are present:
```
Comment (Danziger) [inline app/src/lib/components/public/CheckoutOverlay/CheckoutOverlay.tsx:495]: Big issue here, just noticed this now. Is there a way we can get the backend to sort these by creation date or expose that as a field?

File: app/src/lib/hooks/useFullPayment.ts (hunk 7 above)
Comment (Danziger) [reply app/src/lib/hooks/useFullPayment.ts:224]: As discussed with Baskar, this is not needed. @Dmckinney821

File: app/src/lib/domain/circle/circle.interfaces.ts (hunk 3 above)
Comment (Danziger) [reply app/src/lib/domain/circle/circle.interfaces.ts:30]: Baskar already explained Crypto payments don't go through Circle, thus the different interface. In any case, this is something to keep in mind. It would make sense to eventually have a consistent API / flow for all payment methods.

File: app/src/lib/hooks/useCreatePaymentMethod.ts (hunk 4 above)
Comment (Danziger) [reply app/src/lib/hooks/useCreatePaymentMethod.ts:91]: Baskar already explained, as stated above.

File: app/src/lib/domain/circle/circle.interfaces.ts (hunk 3 above)
Comment (Dmckinney821) [reply app/src/lib/domain/circle/circle.interfaces.ts:30]: Agreed. We need to scope/plan handling crypto within Mojito Checkout to alleviate any need for 3rd party handling

**Comment by Danziger** [comment]
Everything looks good to me @Dmckinney821, feel free to approve/merge once you review it.

Regarding updating marketplaces or other projects, they need to pass these two URL props with their own domain to `PUICheckout`:

- `coinbaseSuccessURL = "http.../payments/success/?from=coinbase"`.
- `coinbaseErrorURL = "http:.../payments/error/?from=coinbase" `.

File: app/src/components/checkout-component/CheckoutComponent.tsx [hunk 9]
```diff
@@ -87,6 +89,8 @@ export const CheckoutComponent: React.FC<PUICheckoutComponentProps> = ({
     productConfirmationEnabled: false,
     vertexEnabled: true,
     threeDSEnabled: true,
+    coinbaseSuccessURL: "http://localhost:3000/payments/success/?from=coinbase",
```
Comment (Dmckinney821) [inline app/src/components/checkout-component/CheckoutComponent.tsx:92]: Should we just add a comment above stating the obvious but that localhost isn't the 'actual' url you'll want to use

File: app/src/lib/components/public/CheckoutOverlay/CheckoutOverlay.tsx [hunk 10]
```diff
@@ -490,30 +495,53 @@ export const PUICheckoutOverlay: React.FC<PUICheckoutOverlayProps> = ({
...
 
-      let matchingPaymentMethod: SavedPaymentMethod | undefined;
+      return {
+        billingInfo: matchingPaymentMethod?.addressId || "",
+        paymentInfo: matchingPaymentMethod?.id || "",
+        paymentType: matchingPaymentMethod?.type || "",
+        cvv: "",
+      };
 
+      /*
```
Comment (Dmckinney821) [inline app/src/lib/components/public/CheckoutOverlay/CheckoutOverlay.tsx:510]: confirming this should be commented out @Danziger

Comment (Danziger) [reply app/src/lib/components/public/CheckoutOverlay/CheckoutOverlay.tsx]: Removed it, I can always get it back from git if needed.

**Review by Dmckinney821 — APPROVED** [review]
